### ISSUE TITLE
Should wasm code be able to extract the externref from exceptions thrown from js?

### ISSUE THREAD

**Issue by Ms2ger** [WebAssembly/exception-handling#202]
It seems like we would need to expose the "JavaScript exception tag" to JS (as a WebAssembly.Tag object, presumably?), so it can be passed as one of the imports when instantiating a wasm module.

I'll try to take a look at defining this after #201.

CC @aheejin

**Comment by dschuff** [comment]
Based on our previous discussions, I think the conclusion was that we do want to do this. As you said, I think that it would be a WebAssembly.Tag object, and when you import it and catch it, you'd get an externref on the stack which corresponds to the object thrown from JS.

**Comment by bakkot** [comment]
I would like to be able to do this. Specifically, I have some wasm code which calls a JS function inside a (wasm) try/catch, and the JS function throws an arbitrary JS exception (or indeed any value, since JS exceptions are not restricted), and inside the (wasm) catch I want to be able to pass that exception as an externref to a different JS function.

Was there ever progress on this?

I see a comment at https://github.com/WebAssembly/exception-handling/pull/218#issuecomment-1249470094 which seems to imply it's "could be deferred to a future version of wasm". Is that right? Is that tracked somewhere?

**Comment by aheejin** [comment]
https://github.com/WebAssembly/exception-handling/pull/218#issuecomment-1249470094 is about preserving exception identities and not about being able to extract JS exceptions from Wasm side. https://github.com/WebAssembly/exception-handling/pull/218#issuecomment-1249470094 basically says, if JS throws an exception and it passes through Wasm (either by not being caught or being caught and rethrown) the exception object you will get is the same as the one you threw.

I think you can define your own [`WebAssembly.Tag`](https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/Tag) in the JS side using the JS API and specify `parameters` as `[externref]`, create an [`WebAssembly.Exception`](https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/Exception) using the JS API and pass your JS object, import the tag you defined from Wasm, and create a `catch` instruction that can catch the imported tag and extract the `externref` thrown. I don't think there's a toolchain yet that enables this process automatically though.

**Comment by michaelficarra** [comment]
@aheejin The exceptions I am interested in catching are not thrown by code that I control. They are not `WebAssembly.Exception` instances. They are any JavaScript value. I want to catch them and pass the externref to a JavaScript FFI that I've defined for processing.

The current strategy I have to take is wrapping the body of any JS FFI that may throw (due to transitively calling something that may throw) in a `try`/`catch` and then in the `catch` block setting a wasm global with the caught value. Then after every call to one of these FFIs from wasm, I check to see if the global has been set and, if so, process it. This works, but it is much less convenient than having a `catch` in wasm for JS exceptions.

**Comment by aheejin** [comment]
> The current strategy I have to take is wrapping the body of any JS FFI that may throw (due to transitively calling something that may throw) in a `try`/`catch` and then in the `catch` block setting a wasm global with the caught value.

How do you catch a JS exception and extract it, given that you don't have a tag for that imported in wasm?

**Comment by michaelficarra** [comment]
```js
function someJSFFI() {
  try {
    // do some stuff
  } catch (e) {
    wasmInstance.exports.storedError.value = e;
  }
}
```
